drivers/sdhc/emmc_host.c

```diff
@@ -0,0 +1,1337 @@
+
+	return emmc_host_send_cmd(emmc, &emmc_cmd);
+}
+
+static int32_t emmc_send_cmd_data(struct emmc_data *emmc, uint32_t cmd_idx,
+				  struct sdhc_command *cmd, struct sdhc_data *data, bool read)
+{
+	struct emmc_cmd_config emmc_cmd;
+	int ret;
+
+	emmc_cmd.sdhc_cmd = cmd;
+	emmc_cmd.cmd_idx = cmd_idx;
+	emmc_cmd.cmd_type = EMMC_HOST_CMD_NORMAL;
+	emmc_cmd.data_present = true;
+	emmc_cmd.idx_check_en = true;
+	emmc_cmd.crc_check_en = true;
+
+	ret = emmc_init_xfr(emmc, data, read);
+	if (ret) {
+		LOG_ERR("Error on init xfr");
+		return ret;
+	}
+
+	ret = emmc_host_send_cmd(emmc, &emmc_cmd);
+	if (ret) {
+		return ret;
+	}
+
+#if !defined(CONFIG_EMMC_HOST_DMA)
+	if (read) {
+		ret = read_data_port(emmc, data);
+	} else {
+		ret = write_data_port(emmc, data);
+	}
+#else
+	ret = wait_xfr_complete(emmc, data->timeout_ms);
+#endif
+
+	return ret;
+}
+
+static int emmc_xfr(struct emmc_data *emmc, struct sdhc_command *cmd, struct sdhc_data *data,
+		    bool read)
+{
+	int ret;
+	struct emmc_cmd_config emmc_cmd;
+
+	ret = emmc_init_xfr(emmc, data, read);
+	if (ret) {
+		LOG_ERR("error emmc init xfr");
+		return ret;
+	}
+	emmc_cmd.sdhc_cmd = cmd;
+	emmc_cmd.cmd_type = EMMC_HOST_CMD_NORMAL;
+	emmc_cmd.data_present = true;
+	emmc_cmd.idx_check_en = true;
+	emmc_cmd.crc_check_en = true;
+
+	k_event_clear(&emmc->irq_event, EMMC_HOST_XFER_COMPLETE);
+	k_event_clear(&emmc->irq_event, read ? EMMC_HOST_BUF_RD_READY : EMMC_HOST_BUF_WR_READY);
+
+	if (data->blocks > 1) {
+		emmc_cmd.cmd_idx = read ? SD_READ_MULTIPLE_BLOCK : SD_WRITE_MULTIPLE_BLOCK;
+		ret = emmc_host_send_cmd(emmc, &emmc_cmd);
+	} else {
+		emmc_cmd.cmd_idx = read ? SD_READ_SINGLE_BLOCK : SD_WRITE_SINGLE_BLOCK;
+		ret = emmc_host_send_cmd(emmc, &emmc_cmd);
+	}
+
+	if (ret) {
+		return ret;
+	}
+
+#if !defined(CONFIG_EMMC_HOST_DMA)
+	if (read) {
+		ret = read_data_port(emmc, data);
+	} else {
+		ret = write_data_port(emmc, data);
+	}
+#else
+	ret = wait_xfr_complete(emmc, data->timeout_ms);
+#endif
+
+#if !defined(CONFIG_EMMC_HOST_AUTO_STOP)
+	emmc_stop_transfer(emmc);
+#endif
+	return ret;
+}
+
+static int emmc_request(const struct device *dev, struct sdhc_command *cmd, struct sdhc_data *data)
+{
+	struct emmc_data *emmc = dev->data;
+	int ret;
+
+	LOG_DBG("");
+
+	if (data) {
+		switch (cmd->opcode) {
+		case SD_WRITE_SINGLE_BLOCK:
+		case SD_WRITE_MULTIPLE_BLOCK:
+			LOG_DBG("SD_WRITE_SINGLE_BLOCK");
+			ret = emmc_xfr(emmc, cmd, data, false);
+			break;
+
+		case SD_READ_SINGLE_BLOCK:
+		case SD_READ_MULTIPLE_BLOCK:
+			LOG_DBG("SD_READ_SINGLE_BLOCK");
+			ret = emmc_xfr(emmc, cmd, data, true);
+			break;
+
+		case MMC_SEND_EXT_CSD:
+			LOG_DBG("EMMC_HOST_SEND_EXT_CSD");
+			ret = emmc_send_cmd_data(emmc, MMC_SEND_EXT_CSD, cmd, data, true);
+			break;
+
+		default:
+			ret = emmc_send_cmd_data(emmc, cmd->opcode, cmd, data, true);
+		}
+	} else {
+		ret = emmc_send_cmd_no_data(emmc, cmd->opcode, cmd);
+
+		/*fixme: WA for adding delay after card switch to HS200 mode */
```

Comment:
Just out of curiosity- what is the specific behavior this works around? It seems like this line is intended to induce a delay when calling `sdmmc_wait_ready` here: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/sd/mmc.c#L454

I wonder a few things here:
- should `sdhc_busy` be returning true after the card switches to HS200 mode until the 10 ms delay expires? This way, the `SD_SEND_STATUS` command is not sent until the card is ready for data
- would increasing `CONFIG_SD_DATA_TIMEOUT` work around this issue as well? IE is the card timing out during the call to `sdmmc_wait_ready`?